core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
+    "Byte"    -> ((source: ReadConverterContext) => (col: Int) => source.getByte(col)),
+    "Double"  -> ((source: ReadConverterContext) => (col: Int) => source.getDouble(col)),
+    "Float"   -> ((source: ReadConverterContext) => (col: Int) => source.getFloat(col)),
+    "Integer" -> ((source: ReadConverterContext) => (col: Int) => source.getInt(col)),
+    "Long"    -> ((source: ReadConverterContext) => (col: Int) => source.getLong(col)),
+    "Short"   -> ((source: ReadConverterContext) => (col: Int) => source.getShort(col)),
+    "String"  -> ((source: ReadConverterContext) => (col: Int) => source.getString(col)))
+
+  private def returnOption[X](op: ReadConverterContext => Int => X) = (source: ReadConverterContext) => (col: Int) => opt(op(source)(col))
+
+  type TypeName = String
+
+  private val allExtractors: Map[TypeName, Deserializer] = plainExtractors ++
+    (plainExtractors map {case (key, op) => s"Option[$key]" -> returnOption(op)})
+
+  private val extractorsMap: Map[TypeName, Deserializer] = allExtractors withDefaultValue DefaultDeserializer
+
+  val extractors = extractorsMap compose columnTypeNames
+
+  private def opt[X](op: => Any): Option[X] = try {
+    Option(op.asInstanceOf[X])
+  } catch {
+    case ex: Exception => None
+  }
+  // maps data columns to product components
+  val columnMapping: Map[Int, Int] =
+    if (columnTypeNames.size == 1) Map(0->0) else multicolumnMapping
+
+  def multicolumnMapping: Map[Int, Int] = {
+    try {
+      val mappings = for {
+        i <- columnTypeNames.indices
+        name = colNamesInResult(i)
+        j = colNamesInFrame.indexOf(name)
+      } yield (i, j)
+
+      val bads = mappings collect {
+        case (i, j) if j < 0 =>
+          if (i < colNamesInResult.length) colNamesInResult(i) else s"[[$i]] (column of type ${columnTypeNames(i)}"
+      }
+
+      if (bads.nonEmpty) {
+        throw new scala.IllegalArgumentException(s"Missing columns: ${bads mkString ","}")
+      }
+
+      mappings.toMap
+    }
+  }
+
+  class H2ORDDIterator(val keyName: String, val partIndex: Int) extends H2OChunkIterator[A] {
+
+    private def cell(i: Int) = {
+      val j = columnMapping(i)
+      val ex = extractors(i)(converterCtx)
+      val data = ex(j).asInstanceOf[Object]
+//      println(s"@$i/$j = $data")
+      data
+    }
+
+    def extractRow: Option[Array[AnyRef]] = {
+      val rowOpt = opt {
+        val objects: IndexedSeq[Object] = columnTypeNames.indices map cell
+        val row = objects toArray
+
+//        println(s"${converterCtx.rowIdx} -> ${res mkString ":"}")
+        row
+      }
+      converterCtx.increaseRowIdx()
+      rowOpt
+    }
+
+    private var hd: Option[A] = None
+    private var total = 0
 
-      def next(): A = {
-        val data = new Array[Option[Any]](numCols)
-        // FIXME: this is not perfect since ncols does not need to match number of names
-        (0 until numCols).foreach{ idx =>
-          val value = if (converterCtx.isNA(idx)) None
```

Comment:
Nope, we have to do it explicitly so far. But as @mmalohlava mentioned, it would be good if they do.

I'll create branch based on your dataset branch @vpatryshev with this change and please feel free to integrate it with your branch before we merge this one.